### PR TITLE
Complete rewrite of Bigcommerce API Client.

### PR DESCRIPTION
**Pull request by pedelman**
- Bump major version to 1.0.0
- Introduces a new configuration syntax
- Support for domain objects
- Support for both legacy and oauth authentication
- Add comprehensive examples
- Add complete testing coverage
- Add complete resource coverage from developer documentation
- Add support for parsing server errors
- And much more!

### CODE REVIEW

**Comment by mattolson** [comment]
excited

File: .travis.yml [hunk 1]
```diff
@@ -1,10 +1,14 @@
 language: ruby
+
```
Comment (splittingred) [inline .travis.yml]: :information_desk_person: Also recommend adding `cache: bundler` in here as well, which will also speed up tests.

File: CONTRIBUTING.md [hunk 2]
```diff
@@ -0,0 +1,119 @@
+# Contributing to Bigcommerce
+
+
+We would love too see contributors! You're encouraged to submit [pull requests](https://github.com/bigcommerce/bigcommerce-api-ruby/pulls), [propose features, and discuss issues](https://github.com/bigcommerce/bigcommerce-api-ruby/issues).
```
Comment (mattolson) [inline CONTRIBUTING.md]: typo: `too` vs `to`.

File: README.md [hunk 3]
```diff
@@ -6,104 +6,96 @@
...
-it and submit a changes with your pull request.
+We have introduced a new major version of the API client and it is a complete rewrite (for the better). If you want to see the old version of the API client, please view it here: [Bigcommerce API client v0.x](https://github.com/bigcommerce/bigcommerce-api-ruby/tree/0.x). We recommend that developers upgrade to the latest client, but we will still support our developers who are unable to upgrade.
 
-## Requirements
 
-- Ruby 1.9+
+## Installation
 
-To connect to the API, you need the following credentials:
+Bigcommerce is available to be pulled down from Rubygems:
```
Comment (mattolson) [inline README.md]: Just say `available on Rubygems`

File: README.md [hunk 4]
```diff
@@ -6,104 +6,96 @@
...
+gem 'bigcommerce'
+```
 
-- Secure URL pointing to a Bigcommerce store
-- Username of an authorized admin user of the store
-- API key for the user
+## Getting Started
 
-A valid API key is required to authenticate requests. To grant API access for a user, go to Control Panel > Users > Edit User and make sure that the 'Enable API access?' checkbox is ticked.
+In order to make requests to our API, you must have registered as a developer and have your credentials ready.
```
Comment (mattolson) [inline README.md]: `you must register`

File: README.md [hunk 5]
```diff
@@ -6,104 +6,96 @@
...
 
-## Installation
+We currently have two different types of authentication schemes you can use depending on your use-case.
```
Comment (mattolson) [inline README.md]: remove `types of`

File: README.md [hunk 6]
```diff
@@ -6,104 +6,96 @@
...
-```sh
-gem install bigcommerce
-```
+To develop a custom integration for one store, your app needs to use Basic Authentication.
 
-## Configuration
+[More Information](https://developer.bigcommerce.com/api/legacy/basic-auth)
 
-#### OAuth Authentication
+#### Single Click Apps
```
Comment (mattolson) [inline README.md]: Rename this to "Public Apps" and put it above Private apps.

File: README.md [hunk 7]
```diff
@@ -6,104 +6,96 @@
...
 
-In order to create a new Bigcommerce application, please visit [developer.bigcommerce.com](https://developer.bigcommerce.com). Its free to sign up and only takes a moment. Once you have created your app, you can get a `client_id`.
+Public apps (also known as Single-Click Apps) must use OAuth authentication. This will allow you to write apps which other businesses can install into their stores.
```
Comment (mattolson) [inline README.md]: "Public apps can be submitted to Bigcommerce App Store, allowing other businesses to install it in their Bigcommerce stores."

File: README.md [hunk 8]
```diff
@@ -6,104 +6,96 @@
...
-require 'bigcommerce'
+#### Single Click (Public Apps):
 
-api = Bigcommerce::Api.new({
-  :store_url => "https://store.mybigcommerce.com",
-  :username  => "username",
-  :api_key   => "api_key"
-})
+```rb
+Bigcommerce.configure do |config|
```
Comment (mattolson) [inline README.md]: I would add a comment in each case as to how you would obtain these values.

File: README.md [hunk 9]
```diff
@@ -6,104 +6,96 @@
...
+end
 ```
-Remember that the fields `:ssl_client_cert`, `:ssl_client_key`, `:ssl_ca_file`
-and `:verify_ssl` are all required when enabling SSL certificates.
 
-## Connecting to the store
+__SSL Configuration:__
 
-Once you have authenticated with the OAuth or Legacy credentials, ping the time method to check that your configuration is working and you can connect successfully to the store:
+If you are using your own self-signed certificate, you can pass in the Faraday options directly. Please note this is not at all required and is only for people who know what they are doing.
```
Comment (mattolson) [inline README.md]: `pass in SSL options to Faraday`.

Comment (mattolson) [inline README.md]: `this is not required, but may be useful in special edge cases.`

File: bigcommerce.gemspec [hunk 10]
```diff
@@ -1,59 +1,28 @@
...
-  s.homepage = "http://github.com/bigcommerce/bigcommerce-api-ruby"
-  s.require_paths = ["lib"]
-  s.rubygems_version = "2.1.11"
-  s.summary = "Enables Ruby applications to communicate with the Bigcommerce API"
-  s.test_files = ["spec/big_commerce_spec.rb", "spec/integration/orders_spec.rb", "spec/unit/api_request_spec.rb", "spec/unit/api_spec.rb", "spec/unit/connection_spec.rb", "spec/unit/date_time_spec.rb", "spec/unit/version_spec.rb"]
 
-  if s.respond_to? :specification_version then
-    s.specification_version = 4
+  s.description = 'Bigcommerce API client in ruby. Allows developers to easily\
```
Comment (mattolson) [inline bigcommerce.gemspec]: `Bigcommerce API Ruby client library. Allows developers to easily communicate with the Bigcommerce API, for either CLI tools or public apps. More info: http://developer.bigcommerce.com`

File: bigcommerce.gemspec [hunk 11]
```diff
@@ -1,59 +1,28 @@
...
+  communicate to the Bigcommerce API allowing developers to write CLI tools or\
+  to create single click apps. Find out more at developer.bigcommerce.com'
+  s.summary = 'Bigcommerce API client to communicate with the v2 API.'
```
Comment (mattolson) [inline bigcommerce.gemspec]: `Ruby client library for Bigcommerce v2 API`

File: lib/bigcommerce/request.rb [hunk 12]
```diff
@@ -0,0 +1,81 @@
...
+
+module Bigcommerce
+  class PathBuilder
+    attr_reader :uri
+
+    def initialize(uri)
+      @uri = uri
+    end
+
+    def build(keys = [])
```
Comment (gregory) [inline lib/bigcommerce/request.rb]: This needs documentation.

**Comment by gregory** [comment]
nice job @pedelman !

**Comment by mattolson** [comment]
slow_clap_360 png

**Comment by splittingred** [comment]
wooot

:+1: :+1:

**Pull request merged**